File: app/src/main/java/com/ngengs/android/popularmovies/apps/fragments/GridFragment.java
package com.ngengs.android.popularmovies.apps.fragments;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;
import androidx.annotation.NonNull;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.ngengs.android.popularmovies.apps.R;
import com.ngengs.android.popularmovies.apps.adapters.MovieListAdapter;
import com.ngengs.android.popularmovies.apps.data.MoviesDetail;
import com.ngengs.android.popularmovies.apps.data.MoviesList;
import com.ngengs.android.popularmovies.apps.data.local.MoviesProviderHelper;
import com.ngengs.android.popularmovies.apps.databinding.FragmentGridBinding;
import com.ngengs.android.popularmovies.apps.globals.Values;
import com.ngengs.android.popularmovies.apps.utils.GridSpacesItemDecoration;
import com.ngengs.android.popularmovies.apps.utils.MoviesAPI;
import com.ngengs.android.popularmovies.apps.utils.NetworkHelpers;
import com.ngengs.android.popularmovies.apps.utils.ResourceHelpers;

import org.reactivestreams.Subscriber;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link GridFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link GridFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class GridFragment extends Fragment {
    private static final String TAG = "GridFragment";
    FragmentGridBinding binding;
    private GridLayoutManager layoutManager;
    private GridSpacesItemDecoration layoutDecoration;
    private Snackbar snackbar;
    private MoviesAPI moviesAPI;
    private MovieListAdapter adapter;
    private Disposable disposable;
    private int sortType;
    private int pageTotal = 1;
    private int pageNow = 0;
    private boolean forceRefresh;
    private boolean loading;
    private boolean fromPagination;
    private boolean changeData;
    private boolean processLoadData = true;
    private final Action actionComplete = this::onComplete;
    private Context context;
    private final Consumer<MoviesList> moviesListConsumer = this::onResponse;
    private MoviesProviderHelper moviesProviderHelper;
    private final Consumer<Throwable> errorConsumer = this::onFailure;
    private OnFragmentInteractionListener mListener;

    public GridFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment GridFragment.
     */
    public static GridFragment newInstance(int sort) {
        GridFragment fragment = new GridFragment();
        Bundle args = new Bundle();
        args.putInt("ARGS_SORT_TYPE", sort);
        fragment.setArguments(args);
        return fragment;
    }

    @SuppressWarnings("EmptyMethod")
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            sortType = getArguments().getInt("ARGS_SORT_TYPE", Values.TYPE_POPULAR);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentGridBinding.inflate(inflater, container, false);
        View view = binding.getRoot();
        context = view.getContext();
        moviesProviderHelper = new MoviesProviderHelper(context);
        createLayout(savedInstanceState);
        if (mListener != null) mListener.onAttachHandler();
        initializeViewAction();
        return view;
    }

    private void initializeViewAction() {
        binding.imageTools.setOnClickListener(view -> imageToolsClick());
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
        binding = null;
    }

    public int getSortType() {
        return sortType;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (adapter.getItemCount() > 0) {
            List<MoviesDetail> data = adapter.get();
            outState.putParcelableArrayList("DATA", new ArrayList<>(data));
            outState.putInt("PAGE_NOW", pageNow);
            outState.putInt("PAGE_TOTAL", pageTotal);
            outState.putInt("SORT_TYPE", sortType);
            outState.putBoolean("PROCESS_LOAD_DATA", processLoadData);
        }
    }

    private void doMoviePressed(int position) {
        Log.d(TAG, "doMoviePressed: " + position);
        if (mListener != null) {
            if (disposable != null && !disposable.isDisposed()) {
                disposable.dispose();
                disposable = null;
                onComplete();
            }

            mListener.onFragmentClickMovies(position, adapter.get(position));
        } else throw new UnsupportedOperationException();

    }

    private void doChangeTitle() {
        Log.d(TAG, "doChangeTitle: " + sortType);
        if (mListener != null) {
            mListener.onFragmentChangeTitle(sortType);
        } else throw new UnsupportedOperationException();
    }

    private void createLayout(Bundle savedInstanceState) {
        Log.d(TAG, "createLayout: now");
        loading = false;
        fromPagination = false;
        changeData = true;

        // Make sure all view not visible
        binding.recyclerView.setVisibility(View.GONE);
        binding.progressBar.setVisibility(View.GONE);

        adapter = new MovieListAdapter(context, this::doMoviePressed);


        int gridSpan;
        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE)
            gridSpan = 4;
        else gridSpan = 2;
        layoutManager = new GridLayoutManager(context, gridSpan);
        layoutDecoration = new GridSpacesItemDecoration(gridSpan, getResources().getDimensionPixelSize(R.dimen.grid_spacing));

        binding.recyclerView.setLayoutManager(layoutManager);
        binding.recyclerView.addItemDecoration(layoutDecoration);
        binding.recyclerView.setAdapter(adapter);
        binding.recyclerView.setHasFixedSize(true);
        binding.recyclerView.setNestedScrollingEnabled(false);
        binding.recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @SuppressWarnings("EmptyMethod")
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                int visibleItemCount = layoutManager.getChildCount();
                int totalItemCount = layoutManager.getItemCount();
                int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();
                boolean endHasBeenReached = (firstVisibleItemPosition + visibleItemCount + 5) >= totalItemCount;

                if (!loading && pageNow < pageTotal && endHasBeenReached && !processLoadData) {
                    Log.d(TAG, "onScrolled: CatchData");
                    forceRefresh = false;
                    fromPagination = true;
                    if (sortType == Values.TYPE_POPULAR) getPopularMovies();
                    else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies();
                }
            }
        });

        forceRefresh = false;
        binding.swipeRefresh.setOnRefreshListener(() -> {
            forceRefresh = true;
            pageNow = 0;
            pageTotal = 1;
            fromPagination = false;
            if (sortType == Values.TYPE_POPULAR) getPopularMovies();
            else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies();
            else getFavoriteMovies();
        });

        moviesAPI = NetworkHelpers.provideAPI();

        if (savedInstanceState != null) {
            sortType = savedInstanceState.getInt("SORT_TYPE", Values.TYPE_POPULAR);
            pageNow = savedInstanceState.getInt("PAGE_NOW", 0);
            pageTotal = savedInstanceState.getInt("PAGE_TOTAL", 1);
            List<MoviesDetail> temp = savedInstanceState.getParcelableArrayList("DATA");
            if (temp != null) {
                adapter.clear();
                adapter.add(temp);
                binding.recyclerView.setVisibility(View.VISIBLE);
                binding.progressBar.setVisibility(View.GONE);
                binding.tools.setVisibility(View.GONE);
                doChangeTitle();
            }
            processLoadData = savedInstanceState.getBoolean("PROCESS_LOAD_DATA", false);
            Log.d(TAG, "createLayout: load savedInstanceState: isProcessLoadData value " + processLoadData);
            Log.d(TAG, "createLayout: load savedInstanceState: sortType value " + (sortType == Values.TYPE_POPULAR));
            if (processLoadData) {
                if (pageNow == 0) adapter.clear();
                if (sortType == Values.TYPE_POPULAR) getPopularMovies();
                else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies();
                else getFavoriteMovies();
            }
        } else {
            if (sortType == Values.TYPE_POPULAR) {
                bindOldData();
                Log.d(TAG, "createLayout: catch new data");
                getPopularMovies();
            } else if (sortType == Values.TYPE_HIGH_RATED) {
                bindOldData();
                Log.d(TAG, "createLayout: catch new data");
                getPopularMovies();
            } else getFavoriteMovies();
            doChangeTitle();
        }
    }

    public void updateSpanColumn(int span) {
        if (layoutManager != null) layoutManager.setSpanCount(span);
        else layoutManager = new GridLayoutManager(context, span);
        binding.recyclerView.setLayoutManager(layoutManager);
        binding.recyclerView.removeItemDecoration(layoutDecoration);
        layoutDecoration = new GridSpacesItemDecoration(span, getResources().getDimensionPixelSize(R.dimen.grid_spacing));
        binding.recyclerView.addItemDecoration(layoutDecoration);
    }

    public void scrollToPosition(int position) {
        layoutManager.scrollToPosition(position);
    }

    private void onResponse(@NonNull MoviesList moviesList) {
        if (binding.recyclerView.getVisibility() == View.GONE) binding.recyclerView.setVisibility(View.VISIBLE);

        if (forceRefresh) adapter.clear();


        if (snackbar != null) {
            snackbar.dismiss();
            snackbar = null;
        }
        if (binding.tools.getVisibility() == View.VISIBLE) binding.tools.setVisibility(View.GONE);
        fromPagination = false;
        pageTotal = moviesList.getTotalPage();
        pageNow = moviesList.getPage();
        List<MoviesDetail> movies = moviesList.getMovies();
        if (moviesList.getStatusMessage() == null && movies.size() > 0) {
            adapter.add(movies);
        }
        if (movies.size() == 0) {
            binding.tools.setVisibility(View.VISIBLE);
            binding.imageTools.setImageDrawable(ResourceHelpers.getDrawable(context, R.drawable.ic_movie_white));
            binding.textTools.setText(R.string.data_empty);
        }
        Log.d(TAG, "onResponse: pageNow: " + pageNow + " pageTotal: " + pageTotal);
        Log.d(TAG, "onResponse: finish Response");
    }

    private void onComplete() {
        if (binding.progressBar.getVisibility() == View.VISIBLE) binding.progressBar.setVisibility(View.GONE);
        if (binding.swipeRefresh.isRefreshing()) binding.swipeRefresh.setRefreshing(false);
        if (forceRefresh) forceRefresh = false;

        processLoadData = false;
        changeData = false;
        loading = false;
        Log.d(TAG, "onComplete: finish complete");
    }

    private void onFailure(@NonNull Throwable t) {
        if (binding.progressBar.getVisibility() == View.VISIBLE) binding.progressBar.setVisibility(View.GONE);
        if (binding.recyclerView.getVisibility() == View.VISIBLE && adapter.getItemCount() == 0)
            binding.recyclerView.setVisibility(View.GONE);

        if (fromPagination) {
            binding.imageTools.setImageDrawable(ResourceHelpers.getDrawable(context, R.drawable.ic_refresh_white));
            binding.textTools.setText(R.string.error_next_page);
            binding.tools.setVisibility(View.VISIBLE);
        } else if (adapter.getItemCount() == 0) {
            binding.imageTools.setImageDrawable(ResourceHelpers.getDrawable(context, R.drawable.ic_cloud_off_white));
            binding.textTools.setText(R.string.error_no_connection);
            binding.tools.setVisibility(View.VISIBLE);
        }

        if (snackbar != null) snackbar.dismiss();
        snackbar = Snackbar.make(binding.textTools, R.string.error_cant_get_data_check_connection, BaseTransientBottomBar.LENGTH_INDEFINITE);
        snackbar.setAction(R.string.retry, v -> {
            if (sortType == Values.TYPE_POPULAR) getPopularMovies();
            else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies();
            else getFavoriteMovies();
        });
        snackbar.show();
        Log.e(TAG, "onFailure: ", t);
    }

    @SuppressWarnings("EmptyMethod")
    @Override
    public void onPause() {
        super.onPause();
    }

    void imageToolsClick() {
        if (sortType == Values.TYPE_POPULAR) getPopularMovies();
        else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies();
        else getFavoriteMovies();
    }

    private void getPopularMovies() {
        Log.d(TAG, "getPopularMovies: pageNow: " + pageNow + " pageTotal: " + pageTotal);
        if (moviesAPI != null && pageNow < pageTotal) {
            Log.d(TAG, "getPopularMovies: now. page: " + pageNow);
            loading = true;
            processLoadData = true;
            if (!forceRefresh || changeData) binding.progressBar.setVisibility(View.VISIBLE);
            binding.tools.setVisibility(View.GONE);

            Log.d(TAG, "getPopularMovies: page: " + pageNow);
            disposable = moviesAPI.listMoviesPopular(pageNow + 1)
                    .subscribeOn(Schedulers.io())
                    .doOnNext(moviesList -> {
                        Log.d(TAG, "getPopularMovies: page: " + pageNow);
                        moviesProviderHelper.saveMovies(moviesList.getMovies());
                        if (pageNow == 0) {
                            moviesProviderHelper.deletePopular();
                            moviesProviderHelper.savePopular(moviesList.getMovies());
                        }
                    })
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(moviesListConsumer, errorConsumer, actionComplete);
        }
    }

    private void getTopRatedMovies() {
        if (moviesAPI != null && pageNow < pageTotal) {
            Log.d(TAG, "getTopRatedMovies: now. page: " + pageNow);
            loading = true;
            processLoadData = true;
            if (!forceRefresh || changeData) binding.progressBar.setVisibility(View.VISIBLE);
            binding.tools.setVisibility(View.GONE);

            disposable = moviesAPI.listMoviesTopRated(pageNow + 1)
                    .subscribeOn(Schedulers.io())
                    .doOnNext(moviesList -> {
                        Log.d(TAG, "getTopRatedMovies: page: " + pageNow);
                        moviesProviderHelper.saveMovies(moviesList.getMovies());
                        if (pageNow == 0) {
                            moviesProviderHelper.deleteTopRated();
                            moviesProviderHelper.saveTopRated(moviesList.getMovies());
                        }
                    })
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(moviesListConsumer, errorConsumer, actionComplete);
        }
    }

    private void getFavoriteMovies() {
        Log.d(TAG, "getFavoriteMovies: now");
        loading = true;
        processLoadData = true;
        if (changeData) binding.progressBar.setVisibility(View.VISIBLE);
        binding.tools.setVisibility(View.GONE);

        disposable = Observable.fromPublisher(new Flowable<MoviesList>() {
            @Override
            protected void subscribeActual(Subscriber<? super MoviesList> s) {
                MoviesList movieList = moviesProviderHelper.getFavorites();
                if (movieList != null) s.onNext(movieList);
                else s.onError(null);
                s.onComplete();
            }
        })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(moviesListConsumer, errorConsumer, actionComplete);
    }

    public void changeType(int sortType) {
        this.sortType = sortType;
        forceRefresh = true;
        pageNow = 0;
        pageTotal = 1;
        loading = false;
        changeData = true;
        doChangeTitle();
        if (disposable != null && !disposable.isDisposed()) disposable.dispose();
        if (this.sortType == Values.TYPE_POPULAR) {
            bindOldData();
            Log.d(TAG, "createLayout: catch new data");
            getPopularMovies();
        } else if (this.sortType == Values.TYPE_HIGH_RATED) {
            bindOldData();
            Log.d(TAG, "createLayout: catch new data");
            getTopRatedMovies();
        } else getFavoriteMovies();
    }

    private void bindOldData() {
        MoviesList temp = null;
        // Catch Offline data
        Log.d(TAG, "createLayout: catch old data");
        if (sortType == Values.TYPE_POPULAR) temp = moviesProviderHelper.getPopular();
        else if (sortType == Values.TYPE_HIGH_RATED) temp = moviesProviderHelper.getTopRated();

        if (temp != null) onResponse(temp);
    }

    public void addMovies(MoviesDetail item) {
        adapter.add(item);
    }

    public void removeMovies(MoviesDetail item) {
        adapter.deleteById(item.getId());
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        void onFragmentClickMovies(int position, MoviesDetail data);

        void onFragmentChangeTitle(int sortType);

        void onAttachHandler();
    }

}

File: app/src/main/java/com/ngengs/android/popularmovies/apps/fragments/GridFragment.kt
package com.ngengs.android.popularmovies.apps.fragments

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.ngengs.android.popularmovies.apps.R
import com.ngengs.android.popularmovies.apps.adapters.MovieListAdapter
import com.ngengs.android.popularmovies.apps.data.local.MoviesDatabaseHelper
import com.ngengs.android.popularmovies.apps.data.remote.MoviesDetail
import com.ngengs.android.popularmovies.apps.data.remote.MoviesList
import com.ngengs.android.popularmovies.apps.databinding.FragmentGridBinding
import com.ngengs.android.popularmovies.apps.fragments.GridFragment.OnFragmentInteractionListener
import com.ngengs.android.popularmovies.apps.globals.Values
import com.ngengs.android.popularmovies.apps.utils.GridSpacesItemDecoration
import com.ngengs.android.popularmovies.apps.utils.ResourceHelpers.getDrawable
import com.ngengs.android.popularmovies.apps.utils.networks.MoviesAPI
import com.ngengs.android.popularmovies.apps.utils.networks.NetworkHelpers
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers


/**
 * A simple [Fragment] subclass.
 * Activities that contain this fragment must implement the
 * [OnFragmentInteractionListener] interface
 * to handle interaction events.
 * Use the [GridFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class GridFragment : Fragment() {
    private var _binding: FragmentGridBinding? = null
    private val binding: FragmentGridBinding get() = _binding!!
    private lateinit var layoutManager: GridLayoutManager
    private lateinit var layoutDecoration: GridSpacesItemDecoration
    private var snackbar: Snackbar? = null
    private lateinit var moviesAPI: MoviesAPI
    private lateinit var adapter: MovieListAdapter
    private var disposable: CompositeDisposable = CompositeDisposable()
    private var sortType = 0
    private var pageTotal = 1
    private var pageNow = 0
    private var forceRefresh = false
    private var loading = false
    private var fromPagination = false
    private var changeData = false
    private var processLoadData = true
    private val actionComplete = Action { this.onComplete() }
    private val moviesListConsumer =
        Consumer { moviesList: MoviesList -> this.onResponse(moviesList) }
    private lateinit var moviesDB: MoviesDatabaseHelper
    private val errorConsumer = Consumer { t: Throwable -> onFailure(t) }
    private var mListener: OnFragmentInteractionListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let { sortType = it.getInt(ARGS_SORT_TYPE, Values.TYPE_POPULAR) }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        _binding = FragmentGridBinding.inflate(inflater, container, false)
        val view: View = binding.root
        moviesDB = MoviesDatabaseHelper(requireContext())
        createLayout(savedInstanceState)
        mListener?.onAttachHandler()
        initializeViewAction()
        return view
    }

    private fun initializeViewAction() {
        binding.imageTools.setOnClickListener { imageToolsClick() }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException("$context must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
        if (!disposable.isDisposed) disposable.dispose()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (!disposable.isDisposed) disposable.dispose()
        _binding = null
    }

    fun getSortType(): Int = sortType

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (adapter.itemCount > 0) {
            val data = adapter.get()
            outState.putParcelableArrayList("DATA", ArrayList(data))
            outState.putInt("PAGE_NOW", pageNow)
            outState.putInt("PAGE_TOTAL", pageTotal)
            outState.putInt("SORT_TYPE", sortType)
            outState.putBoolean("PROCESS_LOAD_DATA", processLoadData)
        }
    }

    private fun doMoviePressed(position: Int) {
        Log.d(TAG, "doMoviePressed: $position")
        mListener?.let { listener ->
            if (!disposable.isDisposed) disposable.dispose()
            onComplete()
            val dataPressed = adapter.get(position)
                ?: throw IllegalAccessException("Can't press unsupported movie")
            listener.onFragmentClickMovies(position, dataPressed)

        } ?: throw UnsupportedOperationException()
    }

    private fun doChangeTitle() {
        Log.d(TAG, "doChangeTitle: $sortType")
        mListener?.onFragmentChangeTitle(sortType) ?: throw UnsupportedOperationException()
    }

    private fun createLayout(savedInstanceState: Bundle?) {
        Log.d(TAG, "createLayout: now")
        loading = false
        fromPagination = false
        changeData = true

        // Make sure all view not visible
        binding.recyclerView.visibility = View.GONE
        binding.progressBar.visibility = View.GONE
        adapter = MovieListAdapter(
            requireContext(),
            object : MovieListAdapter.ClickListener {
                override fun onClickListener(position: Int) {
                    doMoviePressed(position)
                }
            }
        )
        val gridSpan: Int =
            if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) 4 else 2
        layoutManager = GridLayoutManager(context, gridSpan)
        layoutDecoration = GridSpacesItemDecoration(
            gridSpan,
            resources.getDimensionPixelSize(R.dimen.grid_spacing)
        )
        binding.recyclerView.layoutManager = layoutManager
        binding.recyclerView.addItemDecoration(layoutDecoration)
        binding.recyclerView.adapter = adapter
        binding.recyclerView.setHasFixedSize(true)
        binding.recyclerView.isNestedScrollingEnabled = false
        binding.recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val visibleItemCount = layoutManager.childCount
                val totalItemCount = layoutManager.itemCount
                val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()
                val endHasBeenReached =
                    firstVisibleItemPosition + visibleItemCount + 5 >= totalItemCount
                if (!loading && pageNow < pageTotal && endHasBeenReached && !processLoadData) {
                    Log.d(TAG, "onScrolled: CatchData")
                    forceRefresh = false
                    fromPagination = true
                    if (sortType == Values.TYPE_POPULAR) getPopularMovies() else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies()
                }
            }
        })
        forceRefresh = false
        binding.swipeRefresh.setOnRefreshListener {
            forceRefresh = true
            pageNow = 0
            pageTotal = 1
            fromPagination = false
            if (sortType == Values.TYPE_POPULAR) getPopularMovies() else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies() else getFavoriteMovies()
        }
        moviesAPI = NetworkHelpers.provideAPI(requireActivity())
        if (savedInstanceState != null) {
            sortType = savedInstanceState.getInt("SORT_TYPE", Values.TYPE_POPULAR)
            pageNow = savedInstanceState.getInt("PAGE_NOW", 0)
            pageTotal = savedInstanceState.getInt("PAGE_TOTAL", 1)
            val temp: List<MoviesDetail>? = savedInstanceState.getParcelableArrayList("DATA")
            if (temp != null) {
                adapter.clear()
                adapter.add(temp)
                binding.recyclerView.visibility = View.VISIBLE
                binding.progressBar.visibility = View.GONE
                binding.tools.visibility = View.GONE
                doChangeTitle()
            }
            processLoadData = savedInstanceState.getBoolean("PROCESS_LOAD_DATA", false)
            Log.d(
                TAG,
                "createLayout: load savedInstanceState: isProcessLoadData value $processLoadData"
            )
            Log.d(
                TAG,
                "createLayout: load savedInstanceState: sortType value " + (sortType == Values.TYPE_POPULAR)
            )
            if (processLoadData) {
                if (pageNow == 0) adapter.clear()
                if (sortType == Values.TYPE_POPULAR) getPopularMovies() else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies() else getFavoriteMovies()
            }
        } else {
            when (sortType) {
                Values.TYPE_POPULAR -> {
                    bindOldData()
                    Log.d(TAG, "createLayout: catch new data")
                    getPopularMovies()
                }
                Values.TYPE_HIGH_RATED -> {
                    bindOldData()
                    Log.d(TAG, "createLayout: catch new data")
                    getPopularMovies()
                }
                else -> getFavoriteMovies()
            }
            doChangeTitle()
        }
    }

    fun updateSpanColumn(span: Int) {
        layoutManager.spanCount = span
        binding.recyclerView.layoutManager = layoutManager
        binding.recyclerView.removeItemDecoration(layoutDecoration)
        layoutDecoration =
            GridSpacesItemDecoration(span, resources.getDimensionPixelSize(R.dimen.grid_spacing))
        binding.recyclerView.addItemDecoration(layoutDecoration)
    }

    fun scrollToPosition(position: Int) {
        layoutManager.scrollToPosition(position)
    }

    private fun onResponse(moviesList: MoviesList) {
        if (binding.recyclerView.visibility == View.GONE) {
            binding.recyclerView.visibility = View.VISIBLE
        }
        if (forceRefresh) adapter.clear()
        snackbar?.dismiss()
        snackbar = null
        if (binding.tools.visibility == View.VISIBLE) binding.tools.visibility = View.GONE
        fromPagination = false
        pageTotal = moviesList.totalPage
        pageNow = moviesList.page
        val movies = moviesList.movies
        if (moviesList.statusMessage == null && movies.isNotEmpty()) {
            adapter.add(movies)
        }
        if (movies.isEmpty()) {
            binding.tools.visibility = View.VISIBLE
            binding.imageTools.setImageDrawable(
                getDrawable(
                    requireContext(),
                    R.drawable.ic_movie_white
                )
            )
            binding.textTools.setText(R.string.data_empty)
        }
        Log.d(TAG, "onResponse: pageNow: $pageNow pageTotal: $pageTotal")
        Log.d(TAG, "onResponse: finish Response")
    }

    private fun onComplete() {
        if (binding.progressBar.visibility == View.VISIBLE) binding.progressBar.visibility =
            View.GONE
        if (binding.swipeRefresh.isRefreshing) binding.swipeRefresh.isRefreshing = false
        if (forceRefresh) forceRefresh = false
        processLoadData = false
        changeData = false
        loading = false
        Log.d(TAG, "onComplete: finish complete")
    }

    private fun onFailure(t: Throwable) {
        Log.e(TAG, "onFailure: ", t)
        if (binding.progressBar.visibility == View.VISIBLE) binding.progressBar.visibility =
            View.GONE
        if (binding.recyclerView.visibility == View.VISIBLE && adapter.itemCount == 0) binding.recyclerView.visibility =
            View.GONE
        if (fromPagination) {
            binding.imageTools.setImageDrawable(
                getDrawable(
                    requireContext(),
                    R.drawable.ic_refresh_white
                )
            )
            binding.textTools.setText(R.string.error_next_page)
            binding.tools.visibility = View.VISIBLE
        } else if (adapter.itemCount == 0 || pageNow == 0) {
            adapter.clear()
            binding.imageTools.setImageDrawable(
                getDrawable(requireContext(), R.drawable.ic_cloud_off_white)
            )
            binding.textTools.setText(R.string.error_no_connection)
            binding.tools.visibility = View.VISIBLE
        }
        if (snackbar != null) snackbar?.dismiss()
        snackbar = Snackbar.make(
            binding.textTools,
            R.string.error_cant_get_data_check_connection,
            BaseTransientBottomBar.LENGTH_INDEFINITE
        )
        snackbar?.setAction(R.string.retry) { if (sortType == Values.TYPE_POPULAR) getPopularMovies() else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies() else getFavoriteMovies() }
        snackbar?.show()
    }

    private fun imageToolsClick() {
        if (sortType == Values.TYPE_POPULAR) getPopularMovies() else if (sortType == Values.TYPE_HIGH_RATED) getTopRatedMovies() else getFavoriteMovies()
    }

    private fun getPopularMovies() {
        Log.d(TAG, "getPopularMovies: pageNow: $pageNow pageTotal: $pageTotal")
        fetchServerData(
            moviesAPI.listMoviesPopular(pageNow + 1)
        ) { listOf(moviesDB.deletePopular(), moviesDB.savePopular(it.movies)) }
    }

    private fun getTopRatedMovies() {
        Log.d(TAG, "getTopRatedMovies: pageNow: $pageNow pageTotal: $pageTotal")
        fetchServerData(
            moviesAPI.listMoviesTopRated(pageNow + 1)
        ) { listOf(moviesDB.deleteTopRated(), moviesDB.saveTopRated(it.movies)) }
    }

    private fun fetchServerData(moviesApiTarget: Observable<MoviesList>, saveCategoryJob: (MoviesList) -> Collection<Completable>) {
        if (pageNow < pageTotal) {
            Log.d(TAG, "fetchServerData: now. page: $pageNow")
            loading = true
            processLoadData = true
            if (!forceRefresh || changeData) binding.progressBar.visibility = View.VISIBLE
            binding.tools.visibility = View.GONE
            Log.d(TAG, "fetchServerData: page: $pageNow")
            disposable.add(moviesApiTarget.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext {
                    Log.d(TAG, "fetchServerData: save data page: $pageNow")
                    val saveJob = mutableListOf(moviesDB.saveMovies(it.movies))
                    if (pageNow == 0) {
                        saveJob.addAll(saveCategoryJob(it))
                    }
                    disposable.add(Completable.merge(saveJob)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({
                            Log.d(TAG, "fetchServerData: success")
                        }) { throwable -> Log.d(TAG, "fetchServerData: error", throwable) }
                    )
                }
                .subscribe(moviesListConsumer, errorConsumer, actionComplete)
            )
        }
    }

    private fun getFavoriteMovies() {
        Log.d(TAG, "getFavoriteMovies: now")
        loading = true
        processLoadData = true
        if (changeData) binding.progressBar.visibility = View.VISIBLE
        binding.tools.visibility = View.GONE
        disposable.add(
            moviesDB.getFavorites()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally(actionComplete)
                .subscribe(moviesListConsumer, errorConsumer)
        )
    }

    fun changeType(sortType: Int) {
        Log.d(TAG, "changeType")
        this.sortType = sortType
        forceRefresh = true
        pageNow = 0
        pageTotal = 1
        loading = false
        changeData = true
        doChangeTitle()
//        if (!disposable.isDisposed) disposable.dispose()
        when (this.sortType) {
            Values.TYPE_POPULAR -> {
                bindOldData()
                Log.d(TAG, "changeType: catch new data")
                getPopularMovies()
            }
            Values.TYPE_HIGH_RATED -> {
                bindOldData()
                Log.d(TAG, "changeType: catch new data")
                getTopRatedMovies()
            }
            else -> getFavoriteMovies()
        }
    }

    private fun bindOldData() {
        // Catch Offline data
        Log.d(TAG, "bindOldData: catch old data")
        val temp = when (sortType) {
            Values.TYPE_POPULAR -> moviesDB.getPopular()
            Values.TYPE_HIGH_RATED -> moviesDB.getTopRated()
            else -> null
        }
        temp?.let { tempJob ->
            disposable.add(tempJob.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    Log.d(TAG, "bindOldData: finishOld Data")
                    it?.let { onResponse(it) }
                }) { Log.d(TAG, "bindOldData: error", it) }
            )
        }
    }

    fun addMovies(item: MoviesDetail) {
        adapter.add(item)
    }

    fun removeMovies(item: MoviesDetail) {
        adapter.deleteById(item.id)
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     *
     * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
     */
    interface OnFragmentInteractionListener {
        fun onFragmentClickMovies(position: Int, data: MoviesDetail)
        fun onFragmentChangeTitle(sortType: Int)
        fun onAttachHandler()
    }

    companion object {
        private const val TAG = "GridFragment"
        private const val ARGS_SORT_TYPE = "ARGS_SORT_TYPE"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param sort Type of sort ([Values.TYPE_POPULAR], [Values.TYPE_HIGH_RATED], [Values.TYPE_FAVORITE]])
         * @return A new instance of fragment GridFragment.
         */
        fun newInstance(sort: Int): GridFragment = GridFragment().apply {
            val args = Bundle().apply { putInt(ARGS_SORT_TYPE, sort) }
            arguments = args
        }
    }
}